### DISCUSSION TITLE
measuring mains frequency

### DISCUSSION THREAD

**Discussion by creativtransfer** [arendst/Tasmota#14242] [Q&A] [answered]
How can I measure mains frequency with tasmota?
Since germany has reduced its power generation capacities by 9 Gigawatt on new years eve by switching off a significant number of power plants, I'd like to keep an eye on mains frequency with tasmota - ideally down to three decimal places. 

Is there an easy way to measure if you have some mains outlet switch such as Blitzwolf BW-SH6, gosund sp111, sonoff basic or similar device?

**Comment by barbudor** [comment, marked as answer]
Measuring frequency at 0.002% is meaning less
And I don't see the point of measuring the frequency in relation of stopping power plants.
If you power grid doesn't have enough resource, they will just cut off some places. Frequency 0.0000Hz

**Reply by creativtransfer** [reply]
Cher barbudor, c'est peut-être inutile pour toi.

For all others who might not know yet: The frequency gives details about the state of the european power grid:

The grid frequency in the synchronous grid of Continental Europe has a nominal value of 50.0 Hz. In the short term a change of the nominal value by ± 10 mHz is possible to keep the gridtime of the network low.

If the deviation from the nominal value exceeds ±10 mHz, then the primary control is activated. The 10 mHz is the allowable measurement error of 10 mHz to get a dead zone which prevents primary control from delivering power with the false sign.

The current set-value of the mains frequency is 50 Hz.

From ±10 mHz to ±200 mHz the primary control is activated proportional from 0 % to 100 %. Long-term maximum deviations of ±180 mHz are allowed, for short periodes even ±200 mHz are ok. The allowed mains frequency range in normal operation is thus obtained at 49.8 Hz to 50.2 Hz. (ENTSO-E Operation Handbook Load-Frequency Control and Performance)

In case of the breakdown of generation capacity or large consumers, short term deviations until 800 mHz are allowed (49.200 Hz to 50.800 Hz). With higher deviations a massive grid failure is very likely. Then the first step is dropping some consumers or producers (load shedding), if even this fails, then the network ceased operations (blackout) and the network built up afterwards. 


See e.g. https://www.mainsfrequency.com/ and https://www.mainsfrequency.com/frequ_info_en.php

**Reply by sfromis** [reply]
Measuring the mains frequency (with 3 decimals) is already a feature of some energy monitoring devices supported by Tasmota, like Shelly 2.5 and Sonoff DualR3. However, seeing 50.061 Hz and 50.034 Hz at the same time leads me to the observation that the precision of change may be better than the accuracy for current value. As the average for each 24-hour period on the Shelly 2.5 is not entirely stable either, my assumption is that the deviation may be due to limitations of the hardware. Of course, the grid will do a lot to keep the number of cycles very stable to average 50 Hz, regulating to even out any temporary drift.

Reviewing the last month, on a Shelly 2.5 (which may have instability, deviations and so on), I'm seeing fluctuations between 49.9 and 50.1 Hz, with a few outliers which I tend to assume is part of the limited accuracy expectable from a device where frequency monitoring is a very secondary feature. Does not appear to be changed since 2022 arrived. While the standard deviation for today is a bit higher than full-day periods from december, this does not tell anything real when it is not a full 24-hour period, and only one day. However, the part of Denmark I live in is part of the Nordic synchronous grid, not tied to continental Europe.

**Reply by creativtransfer** [reply]
Thank you for your detailed reply! 

Do you have hands on experience with the Sonoff DualR3 as well? 
Since I am thinking about getting one of these then it would be good to know if the sonoff has the same issues as the Shelly.
Is there a page similar to https://www.netzfrequenzmessung.de/ for the danish grid? If so, have you compared it with your live data?

**Reply by sfromis** [reply]
I had forgotten to set `FreqRes 3` till today on my Sonoff Dual R3 switches. My impression is that it is not really different from the Shelly 2.5 wrt frequency measurement.

However I can add that my short summarization of data may not be very good for checking details of frequency and accuracy of cheap IoT meters. The thing is that I got the frequency from a sample at every TelePeriod (5 minutes), and when looking at current numbers, I can see them "pulsating" a bit up and down, which looks a lot like ongoing adjustments to match shifting loads and production (including lots of windmill power). This pattern of a "pulse" also makes it impractical to really compare the meters by looking at the numbers "at the same time", as they just may have the sampling interval (1 second) hit differently in the waves.

More generally, as counting the sine wave peaks should be a trivially simple thing in an energy monitoring chip, I do not see why one would be better than the other at counting. Any deviation could be down to the real time clock in these small MCU chip not being rock stable (Tasmota will synchronize with a NTP server hourly to keep good time). My experience is that there's not much drift, so the frequency numbers are probably quite decent.

I'm not aware of where I can see "official" data for the Nordic grid, but I've not looked a lot either.

Anyway, if you want to check for signs of badness, the important thing is not to get an exact frequency, but more to see if there are major dips suggesting trouble.

**Reply by creativtransfer** [reply]
> Anyway, if you want to check for signs of badness, the important thing is not to get an exact frequency, but more to see if there are major dips suggesting trouble.

That's what I am after :) So you basically say, it doesn't matter which of the aforementioned devices I am going to buy. Or does one of them have an advantage which makes it superior?

Without having investigated deeper yet, I am thinking about implementing a tasmota rule which will monitor the frequency and in case it drops below a certain value will send some sort of notification or alert.

**Reply by sfromis** [reply]
My guess is indeed that both devices should be about the same for checking grid frequency. 

If you consider doing more detailed analysis, the Sonoff DualR3 does have the advantage of having the ESP32 chip instead of the older ESP8266. This gives room for a "real programming language", Berry, which is a nice language for doing processing, like checking values at short intervals. While a "classic" Tasmota rule can work fine for just checking if the frequency dropped "too low", when it gets more involved than that, Berry is much better.

Note: To flash alternate firmware, like Tasmota, to the Sonoff DualR3 you need to open the case and connect wires to small soldering pads. With Shelly 2.5, OTA flashing is available, and there is also a small row allowing pins to be connected without opening up the case.

**Reply by creativtransfer** [reply]
Thank you. Then I'll get the sonoff since the ESP32 and will have a look at Berry.

**Comment by digiblur** [comment]
In cases of power grid fluctuations during extreme heat and cold events where they send out conversation messages I've always seen the frequency stay locked in.   It was the voltage that was dropping.

**Reply by creativtransfer** [reply]
That's an interesting point. Where do you live?

**Comment by Jambola77** [comment]
You can see the frequency for Germany (Europe?) here: https://www.netzfrequenz.info/aktuelle-netzfrequenz-full
